### PR TITLE
SONAR-10368 Accelerate PR analysis like a short living branch and similar issue tracking

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/ComponentDtoToWsComponent.java [hunk 1]
```diff
@@ -61,6 +61,7 @@ private ComponentDtoToWsComponent() {
       .setKey(dto.getKey())
       .setName(dto.name())
       .setQualifier(dto.qualifier());
+    // TODO add pull request id to the response when it's available in ComponentDto
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/ComponentDtoToWsComponent.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fpr-issue-tracking&issues=AWGmMPQEuNQTvAhV6oAQ&open=AWGmMPQEuNQTvAhV6oAQ)

[ISSUE_KEY]: AWGmMPQEuNQTvAhV6oAQ

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SendIssueNotificationsStep.java [hunk 2]
```diff
@@ -185,9 +187,17 @@ public String getDescription() {
...
   private String getBranchName() {
     Branch branch = analysisMetadataHolder.getBranch();
-    return branch.isMain() ? null : branch.getName();
+    return branch.isMain() || branch.getType() == PULL_REQUEST ? null : branch.getName();
+  }
+
+  @CheckForNull
+  private String getPullRequest() {
+    Branch branch = analysisMetadataHolder.getBranch();
+    // TODO rely on the pull request instead of the branch name
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SendIssueNotificationsStep.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fpr-issue-tracking&issues=AWGoKgSrWHTwXubXzY2o&open=AWGoKgSrWHTwXubXzY2o)

[ISSUE_KEY]: AWGoKgSrWHTwXubXzY2o

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/AppAction.java [hunk 3]
```diff
@@ -178,6 +178,7 @@ private void appendComponent(JsonWriter json, ComponentDto component, UserSessio
     if (branch != null) {
       json.prop("branch", branch);
     }
+    //TODO set pullRequest and pullRequestTitle
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/AppAction.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fpr-issue-tracking&issues=AWGoKgU_WHTwXubXzY2p&open=AWGoKgU_WHTwXubXzY2p)

[ISSUE_KEY]: AWGoKgU_WHTwXubXzY2p

File: server/sonar-server/src/main/java/org/sonar/server/issue/notification/IssueChangeNotification.java [hunk 4]
```diff
@@ -54,15 +54,19 @@ public IssueChangeNotification setIssue(DefaultIssue issue) {
   }
 
   public IssueChangeNotification setProject(ComponentDto project) {
-    return setProject(project.getKey(), project.name(), project.getBranch());
+    // TODO set the pull request instead of the branch
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/notification/IssueChangeNotification.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fpr-issue-tracking&issues=AWGoKgVrWHTwXubXzY2q&open=AWGoKgVrWHTwXubXzY2q)

[ISSUE_KEY]: AWGoKgVrWHTwXubXzY2q

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/ComponentDtoToWsComponent.java (hunk 1 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/ComponentDtoToWsComponent.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fpr-issue-tracking&issues=AWGmMPQEuNQTvAhV6oAQ&open=AWGmMPQEuNQTvAhV6oAQ)

[ISSUE_KEY]: AWGmMPQEuNQTvAhV6oAQ

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/AppAction.java (hunk 3 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/AppAction.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fpr-issue-tracking&issues=AWGoKgU_WHTwXubXzY2p&open=AWGoKgU_WHTwXubXzY2p)

[ISSUE_KEY]: AWGoKgU_WHTwXubXzY2p

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/ComponentDtoToWsComponent.java (hunk 1 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/ComponentDtoToWsComponent.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fpr-issue-tracking&issues=AWGmMPQEuNQTvAhV6oAQ&open=AWGmMPQEuNQTvAhV6oAQ)

[ISSUE_KEY]: AWGmMPQEuNQTvAhV6oAQ

File: server/sonar-server/src/main/java/org/sonar/server/issue/notification/IssueChangeNotification.java (hunk 4 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/notification/IssueChangeNotification.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fpr-issue-tracking&issues=AWGoKgVrWHTwXubXzY2q&open=AWGoKgVrWHTwXubXzY2q)

[ISSUE_KEY]: AWGoKgVrWHTwXubXzY2q

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SendIssueNotificationsStep.java (hunk 2 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SendIssueNotificationsStep.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fpr-issue-tracking&issues=AWGoKgSrWHTwXubXzY2o&open=AWGoKgSrWHTwXubXzY2o)

[ISSUE_KEY]: AWGoKgSrWHTwXubXzY2o

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/AppAction.java (hunk 3 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/AppAction.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fpr-issue-tracking&issues=AWGoKgU_WHTwXubXzY2p&open=AWGoKgU_WHTwXubXzY2p)

[ISSUE_KEY]: AWGoKgU_WHTwXubXzY2p

File: server/sonar-server/src/main/java/org/sonar/server/issue/notification/IssueChangeNotification.java (hunk 4 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/notification/IssueChangeNotification.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fpr-issue-tracking&issues=AWGoKgVrWHTwXubXzY2q&open=AWGoKgVrWHTwXubXzY2q)

[ISSUE_KEY]: AWGoKgVrWHTwXubXzY2q

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/ComponentDtoToWsComponent.java (hunk 1 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/ComponentDtoToWsComponent.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fpr-issue-tracking&issues=AWGmMPQEuNQTvAhV6oAQ&open=AWGmMPQEuNQTvAhV6oAQ)

[ISSUE_KEY]: AWGmMPQEuNQTvAhV6oAQ

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SendIssueNotificationsStep.java (hunk 2 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SendIssueNotificationsStep.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fpr-issue-tracking&issues=AWGoKgSrWHTwXubXzY2o&open=AWGoKgSrWHTwXubXzY2o)

[ISSUE_KEY]: AWGoKgSrWHTwXubXzY2o

File: server/sonar-server/src/main/java/org/sonar/server/issue/notification/IssueChangeNotification.java (hunk 4 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/notification/IssueChangeNotification.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fpr-issue-tracking&issues=AWGoKgVrWHTwXubXzY2q&open=AWGoKgVrWHTwXubXzY2q)

[ISSUE_KEY]: AWGoKgVrWHTwXubXzY2q

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/ComponentDtoToWsComponent.java (hunk 1 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/ComponentDtoToWsComponent.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fpr-issue-tracking&issues=AWGmMPQEuNQTvAhV6oAQ&open=AWGmMPQEuNQTvAhV6oAQ)

[ISSUE_KEY]: AWGmMPQEuNQTvAhV6oAQ

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SendIssueNotificationsStep.java (hunk 2 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SendIssueNotificationsStep.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fpr-issue-tracking&issues=AWGoKgSrWHTwXubXzY2o&open=AWGoKgSrWHTwXubXzY2o)

[ISSUE_KEY]: AWGoKgSrWHTwXubXzY2o

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/AppAction.java (hunk 3 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/AppAction.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fpr-issue-tracking&issues=AWGoKgU_WHTwXubXzY2p&open=AWGoKgU_WHTwXubXzY2p)

[ISSUE_KEY]: AWGoKgU_WHTwXubXzY2p

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SendIssueNotificationsStep.java (hunk 2 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SendIssueNotificationsStep.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fpr-issue-tracking&issues=AWGoKgSrWHTwXubXzY2o&open=AWGoKgSrWHTwXubXzY2o)

[ISSUE_KEY]: AWGoKgSrWHTwXubXzY2o

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/ComponentDtoToWsComponent.java (hunk 1 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/ComponentDtoToWsComponent.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fpr-issue-tracking&issues=AWGmMPQEuNQTvAhV6oAQ&open=AWGmMPQEuNQTvAhV6oAQ)

[ISSUE_KEY]: AWGmMPQEuNQTvAhV6oAQ

File: server/sonar-server/src/main/java/org/sonar/server/issue/notification/IssueChangeNotification.java (hunk 4 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/notification/IssueChangeNotification.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fpr-issue-tracking&issues=AWGoKgVrWHTwXubXzY2q&open=AWGoKgVrWHTwXubXzY2q)

[ISSUE_KEY]: AWGoKgVrWHTwXubXzY2q

File: sonar-ws/src/main/java/org/sonarqube/ws/client/users/UsersService.java [hunk 5]
```diff
@@ -74,7 +74,7 @@ public CreateWsResponse create(CreateRequest request) {
         .setParam("login", request.getLogin())
         .setParam("name", request.getName())
         .setParam("password", request.getPassword())
-        .setParam("scmAccount", request.getScmAccount())
+        .setParam("scmAccount", request.getScmAccount() == null ? null : request.getScmAccount())
```
Comment (janos-ss) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/users/UsersService.java]: What's the point of this change?

File: sonar-ws/src/main/java/org/sonarqube/ws/client/users/UsersService.java [hunk 6]
```diff
@@ -74,7 +74,7 @@ public CreateWsResponse create(CreateRequest request) {
...
         .setParam("scmAccounts", request.getScmAccounts() == null ? null : request.getScmAccounts().stream().collect(Collectors.joining(","))),
```
Comment (janos-ss) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/users/UsersService.java]: I see other tests remove `scmAccounts` param in favor of singular `scmAccount`, used with list values.

Also, I don't see how this change is related to the task at hand with pull requests.

File: sonar-ws/src/main/java/org/sonarqube/ws/client/qualitygates/ListRequest.java [hunk 7]
```diff
@@ -19,6 +19,7 @@
  */
 package org.sonarqube.ws.client.qualitygates;
 
+import java.util.List;
```
[the remark this answers is not in the document]
Comment (teryk) [reply sonar-ws/src/main/java/org/sonarqube/ws/client/qualitygates/ListRequest.java]: generated code

File: sonar-ws/src/main/java/org/sonarqube/ws/client/users/UsersService.java (hunk 5 above)
Comment (teryk) [reply sonar-ws/src/main/java/org/sonarqube/ws/client/users/UsersService.java]: no point, I badly fixed the generated code

File: sonar-ws/src/main/java/org/sonarqube/ws/client/users/UsersService.java (hunk 6 above)
Comment (teryk) [reply sonar-ws/src/main/java/org/sonarqube/ws/client/users/UsersService.java]: I did this while updating the WS client, the scmAccounts parameter is deprecated

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/sensor/DefaultSensorContext.java [hunk 8]
```diff
@@ -148,15 +148,15 @@ public NewSymbolTable newSymbolTable() {
...
+    if (branchConfiguration.isShortLivingBranch() || branchConfiguration.isPullRequest()) {
       return NO_OP_NEW_COVERAGE;
     }
     return new DefaultCoverage(sensorStorage);
   }
 
   @Override
   public NewCpdTokens newCpdTokens() {
-    if (analysisMode.isIssues() || branchConfiguration.isShortLivingBranch()) {
+    if (analysisMode.isIssues() || branchConfiguration.isShortLivingBranch() || branchConfiguration.isPullRequest()) {
```
[the remark this answers is not in the document]
Comment (teryk) [reply sonar-scanner-engine/src/main/java/org/sonar/scanner/sensor/DefaultSensorContext.java]: as discussed `isShortOrPullRequest`

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/ComponentDtoToWsComponent.java (hunk 1 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/ComponentDtoToWsComponent.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fpr-issue-tracking&issues=AWGmMPQEuNQTvAhV6oAQ&open=AWGmMPQEuNQTvAhV6oAQ)

[ISSUE_KEY]: AWGmMPQEuNQTvAhV6oAQ

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SendIssueNotificationsStep.java (hunk 2 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SendIssueNotificationsStep.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fpr-issue-tracking&issues=AWGoKgSrWHTwXubXzY2o&open=AWGoKgSrWHTwXubXzY2o)

[ISSUE_KEY]: AWGoKgSrWHTwXubXzY2o

File: server/sonar-server/src/main/java/org/sonar/server/issue/notification/IssueChangeNotification.java (hunk 4 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/notification/IssueChangeNotification.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fpr-issue-tracking&issues=AWGoKgVrWHTwXubXzY2q&open=AWGoKgVrWHTwXubXzY2q)

[ISSUE_KEY]: AWGoKgVrWHTwXubXzY2q

**Pull request merged**